cip/1.accepted/CIP2015-05-13-EXISTS.adoc

```diff
@@ -4,102 +4,188 @@
 
-* Property existence checking
-* Subquery existence checking
+* Restricting https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] to predicate contexts and boolean evaluation, when used in expression contexts
+* Adding two forms of existential subqueries denoted with `EXISTS { ... }`:
+** Full existential subquery based on a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RegularQuery[<RegularQuery>]
+** Simple existential subquery based on a single https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Pattern[<Pattern>] and an optional https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=Where[<Where>] clause
+* Removing `exists( n.prop )` in favor of the https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=NullOperatorExpression[<NullOperatorExpression>], which is a restatement of what has been accepted with https://github.com/opencypher/openCypher/blob/master/cip/1.accepted/CIP2018-10-29-EXISTS-and-IS-NOT-NULL.adoc[CIP2018-10-29 EXISTS and IS NOT NULL]
 
 === Syntax
 
-----
-expression = <current definition of expression>
-           | property exists
-           | subquery exists
-           | simple subquery exists
-           ;
+==== Grammar
 
-property exists = "exists", "(", expression ")"
-                ;
+[source,bnf]
+----
+<Atom> ::=
+    ...
+  | <RelationshipsPattern>
+  | <ExistentialSubquery>
 
-subquery exists = "EXISTS", "{", read only clause, { read only clause }, "}" ;
+<ExistentialSubquery> ::=
+    <SimpleExistentialSubquery>
+  | <FullExistentialSubquery>
 
-simple subquery exists = "EXISTS", "{", simple match, "}" ;
+<SimpleExistentialSubquery> ::=
+  "EXISTS", "{", <SimpleMatch>, "}"
 
-simple match = pattern, { ",", pattern }, [ "WHERE", predicate ] ;
+<FullExistentialSubquery> ::=
+  "EXISTS", "{", <RegularQuery>, "}"
 
-read only clause = match
-                 | unwind
-                 | with
-                 ;
+<SimpleMatch> ::=
+  <Pattern>, [ <Where> ]
 ----
 
+Note that the openCypher grammar does not list <SimpleExistentialSubquery>, <SimpleMatch>, and <FullExistentialSubquery> as separate productions but represents them inline of https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=ExistentialSubquery[<ExistentialSubquery>].
+
+==== Syntax Rules
+
+* A https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall only be contained in sites whose expected type is exactly boolean. Specifically, a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall only be simply contained in
```

Comment:
This is in direct conflict with a previous CLG decision to allow `EXISTS {  }` as a general expression: https://trello.com/c/yUwor3uw/532-allow-exists-as-an-expression

Reply:
This line doesn't say anything about `EXISTS { }`, does it? It talks about the "pattern expression" form, ie `WITH ()-->() AS boolean`-style things.

Later on it says

>All three forms ... are boolean expressions

So I don't really see a conflict?

Reply:
Specifically, what I read from this line is that we cannot do `size( ()-->() )`, because `size()` does not expect a boolean. Hence the other discussion we had in CLG last meeting.

Reply:
Right, I misread that. In that case I retract my comment about the conflict.

On a different note, I think the following wording is confusing.
> A RelationshipsPattern shall only be contained in sites whose expected type is exactly boolean. Specifically, a RelationshipsPattern shall only be simply contained in Where, ...

There are more places than listed here where the expected type is exactly boolean. For example `RETURN NOT ()--()`. We deprecated these use cases and I expect they would be forbidden in this proposal, but the wording is unclear.

Reply:
I don't think we need to list every grammar rule that has exactly the expected type boolean. What do you think about rephrasing like this or similar?

```suggestion
* A https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] shall only be contained in sites whose expected type is exactly boolean. To give the most frequent examples, a https://raw.githack.com/openCypher/openCypher/master/tools/grammar-production-links/grammarLink.html?p=RelationshipsPattern[<RelationshipsPattern>] can be simply contained in
```